src/core/internalMemMgr.cpp

```diff
@@ -563,4 +576,116 @@ uint32 InternalMemMgr::GetReferencesCount()
     return static_cast<uint32>(m_references.NumElements());
 }
 
+Result InternalMemMgr::Map(
+        GpuMemory* pGpuMemory,
+        void** ppData)
+{
+    Util::MutexAuto allocatorLock(&m_allocatorLock); // Ensure thread-safety using the lock
+    PAL_ASSERT(pGpuMemory != nullptr);
+
+    Result result = Result::ErrorInvalidValue;
+    if (pGpuMemory->WasBuddyAllocated())
+    {
+        // Try to find the allocation in the pool list
+        for (auto it = m_poolList.Begin(); it.Get() != nullptr; it.Next())
+        {
+            GpuMemoryPool* pPool = it.Get();
+
+            PAL_ASSERT((pPool->pGpuMemory != nullptr) && (pPool->pBuddyAllocator != nullptr));
+
+            if (pPool->pGpuMemory == pGpuMemory)
+            {
+                if (!pPool->mapped)
+                {
+                    result = pPool->pGpuMemory->Map(&pPool->pData);
+                    if (result != Result::Success) {
+                        --m_numMappedPools;
+                        break;
+                    }
+                    pPool->mapped = true;
+                    ++m_numMappedPools;
+                    CheckMappedPoolLimit();
+                } else if (pPool->refCount == 0) {
+                    // should be in unused list, remove it from there.
+                    for (auto it2 = m_unusedMappedPoolList.Begin(); it2.Get() != nullptr; it2.Next()) {
+                        if (*(it2.Get()) == pPool) {
+                            m_unusedMappedPoolList.Erase(&it2);
+                            break;
+                        }
+                    }
+                }
+                pPool->refCount++;
+                *ppData = pPool->pData;
+                result = Result::Success;
+                break;
+            }
+        }
+
+        // If we didn't find the allocation in the pool list then something went wrong with the allocation scheme
+        PAL_ASSERT(result == Result::Success);
+    }
+    else
+    {
+        result = pGpuMemory->Map(ppData);
+    }
+
+    return result;
+}
+
+Result InternalMemMgr::Unmap(
+        GpuMemory* pGpuMemory)
+{
+    Util::MutexAuto allocatorLock(&m_allocatorLock); // Ensure thread-safety using the lock
```

Comment:
Calling `WasBuddyAllocated()` is fine to do outside the lock, so the lock should only be taken inside the scope of that check.

Reply:
Done